coil-network-core/src/commonMain/kotlin/coil3/network/InFlightRequestStrategy.kt

```diff
@@ -0,0 +1,173 @@
+ * `shouldWaitForSignal`, `currentChannelController`, and `isResponsibleForCleanup`.
+ * - The first coroutine to request a key becomes the `currentChannelController`
+ * and is also marked `isResponsibleForCleanup`. It creates and stores the `InFlightRequest`.
+ * - Subsequent coroutines for the same key are marked `shouldWaitForSignal = true`
+ * and suspend by attempting to receive from the request's channel.
+ *
+ * 3. Baton Passing on Failure:
+ * - If a waiting coroutine successfully receives a `Unit` from the channel, it means
+ * the previous `currentChannelController` failed and passed the "baton." This
+ * waiter then becomes the new `currentChannelController` and `isResponsibleForCleanup`.
+ * - If a waiter receives `null` (channel closed), it means the operation likely
+ * succeeded or failed terminally. The waiter proceeds to execute `block()`
+ * (e.g., to hit a cache) but does not become a controller or responsible for cleanup.
+ *
+ * 4. Block Execution & Error Handling:
+ * - All coroutines eventually execute the provided `block()`.
+ * - If `block()` throws an exception for any coroutine:
+ * That coroutine will attempt to send a `Unit` signal on the channel
+ * (`request.channel.trySend(Unit)`).
+ * If this `trySend` is successful (meaning another waiter received the signal and can retry),
+ * and if the currently failing coroutine *was* `isResponsibleForCleanup`, it relinquishes
+ * this responsibility by setting `isResponsibleForCleanup = false`.
+ *
+ * 5. Cleanup:
+ * - In the `finally` block, if a coroutine is both the `currentChannelController` AND
+ * `isResponsibleForCleanup`, it will close the channel and remove the
+ * `InFlightRequest` from the map. This ensures resources are freed when the
+ * operation succeeds, or when it fails and no other waiter takes over.
+ */
+class DeDupeInFlightRequestStrategy() : InFlightRequestStrategy {
+    private val inFlightRequests = mutableMapOf<String, Channel<Unit>>()
+    private val lock = SynchronizedObject()
+
+    override suspend fun apply(
+        key: String,
+        block: (suspend () -> FetchResult),
+    ): FetchResult {
+        var shouldWaitForSignal = true
+        var currentChannelController = false
+        var isResponsibleForCleanup = false
+
+        val channel = synchronized(lock) {
+            inFlightRequests.getOrPut(key) {
+                shouldWaitForSignal = false
+                currentChannelController = true
+                isResponsibleForCleanup = true
+                Channel()
+            }
+        }
+
+        if (shouldWaitForSignal) {
+            if (channel.receiveCatching().getOrNull() != null) {
+                currentChannelController = true
+                isResponsibleForCleanup = true
+            }
+        }
+
+        try {
+            return block()
+        } catch (ex: Exception) {
+            val successfullyPassedBaton = channel.trySend(Unit).isSuccess
```

Comment:
Will this launch a new request if the first request is cancelled and it falls back to a second observer? I'm wondering if we can create an implementation that refcounts the observers and only cancels the request if the refcount drops to 0.

Reply:
This works like a serial queue until `block()` succeeds/returns. 
If block throws - either cancelation or it fails (404 for example), it will awake exactly one waiter to try again.
If block succeeds, it will close the channel, releasing all waiters to proceed.